Add draft content to CEP-47 Implementaion and deploy sections (#206)
casper-network/docs: source/docs/casper/dapp-dev-guide/tutorials/cep47/deploy.md

Let's dive into CEP-47 contract deployment process.

### Pre-requisites
- Set up your machine as per the [prerequisites](/docs/workflow/setup)
- Ensure you have an [Account](https://casper.network/docs/workflow/setup#setting-up-an-account), the associated key pair,  and the required amount of tokens to perform the deploy
- Ensure your [account](https://testnet.cspr.live/tools/faucet) contains enough `CSPR Tokens` to perform the execution. 
- CSPR Tokens are used to pay for the transactions on the Casper Network. Follow the [transfer tokens](https://casper.network/docs/workflow/token-transfer#2-the-faucet) guide to learn more about token transferring on the Casper Testnet

## Basic Flows of the Deployment
Here are the basic steps to deploy the CEP-47 contract on the Casper Network.
<img src={useBaseUrl("/image/tutorials/cep-47/CEP-47-deploy-flow.png")} alt="cep-47-deploy-flow" width="800"/>

### Casper Repositories in Use

You will be using two Casper repositories for this deployment process.

> -  [CEP-47 Contract Implementation](https://github.com/casper-ecosystem/casper-nft-cep47): This repository contains the implementation of the NFT aka CEP-47 smart contract according to Casper context, required utility Classes, and corresponding test suit to work with the CEP-47 token.

Reviewer: @Buddhika-Pathirathna I don't see the changes Fraser suggested in the latest files, so I am unresolving this conversation. Could you please double-check that the changes took effect?

    - You will be using the `cep47-token.wasm` file from this repository for the deployment. This is the compiled implementation of the CEP-47 contract.
-  [Casper JavaScript Client](https://github.com/casper-network/casper-contracts-js-clients): This repository contains the files which execute the Javascript handle for the CEP-47 contract and other supporting classes to run the client. This contains the JavaScript handle for ERC-20 contract as well. 

>     - You will be executing the `install.ts` file and `usage.ts` from this repository for the deployment

Reviewer 2: Seems like executing `usage.ts` is not covered in this page.

> ### Steps to Perform the Deployment

Author: I would like to keep it as a ToC since the flow is a bit long, so users can get a quick idea about the next sections. Also, if anyone wants to jump to any specific section, they can use the links. What are your thoughts @ipopescu, @GloreenS, and @ACStoneCL?

> ### Steps to Perform the Deployment

Reviewer: I usually like TOCs, but many developers just search for content. We can use the right-hand navigation to jump to different sections. ~~I would be ok either way here. Keep it if you feel strongly :).~~
On second thought, I think we should remove the list. Let's use the right-hand navigation pane as a TOC. We don't have TOCs in the docs, and they usually go at the beginning of a document, so it might be better to keep the flow consistent with other articles.

1. [Preparing the CEP-47 smart contract](./deploy#preparing-the-cep-47-smart-contract)
2. [Getting and preparing node addresses for the deployment](./deploy#getting-and-preparing-node-addresses-for-the-deployment)
3. [Preparing JS Client repository](./deploy#preparing-js-client-repository)
4. [Deploying the contract](./deploy#deploying-the-smart-contract-to-the-network)
5. [Deploying the entry points of the contract](../cep47/events)

> ## Preparing the CEP-47 Smart Contract

Author: I will check on this and adjust - Thanks.

Refer to the [contract preparation](../cep47/prepare) step to prepare the CEP-47 contract for deployment. This step will make the build environment, create the target location and compile the contract to a .wasm file.

- Output from this will be a `WASM file` (Eg:- cep47-token.wasm) which is later used by the JS compiler for the contract deployment

## Getting and Preparing Node Addresses for the Deployment
Follow [getting an ip-address from a testnet peer](../erc20/deploy#getting-an-ip-address-from-a-testnet-peer) guide to get a node address for the deployment. Then you have to do some formatting to the address to include in the .env file.

#### NODE_ADDRESS: 
- IP Address of the node

> -  Use port '7777' instead of '35000' or any other and add "/rpc" to the end of the address 

Reviewer: @Buddhika-Pathirathna, I don't see Fraser's suggestion applied, and I think it should be there. Could you please double check all resolved comments?

>     - (Eg:- 195.201.174.222:7777**/rpc**)

Reviewer: Same here. Please double-check all resolved comments, as I am still seeing the previous formatting.

     
#### EVENT_STREAM_ADDRESS: 
- Address of the event stream
- To get this, replace 7777/rpc with --> '9999/events/main' phrase 
    - (Eg: 195.201.174.222:**9999/events/main**)

## Preparing JS Client Repository
This JavaScript client helps to install the smart contract on the Casper network and proceed with related actions with CEP-47 token contract. We are using the JavaScript client classes to invoke the NFT installation on the network using pre-defined set of environment variables.

This step includes several sub-steps to achieve the execution.

1. [Clone the JS client repository](./deploy#cloning-the-js-client-repository)
2. [Add the environment variables for the execution](./deploy#adding-the-environment-variables)
3. [Arrange the JS packages and resolve their dependencies using NodeJS](./deploy#arrange-the-js-packages-and-resolve-dependencies)

> 4. [Build the JS client code](./deploy#building-the-js-client-code)

Reviewer: Agreed, @Buddhika-Pathirathna

### Cloning the JS Client Repository
Clone the [casper-contracts-js-clients](https://github.com/casper-network/casper-contracts-js-clients) repository using the below command.

```
git clone https://github.com/casper-network/casper-contracts-js-clients.git
```

### Adding the Environment Variables 
Environment file (.env) contains the individual user environment variables required to execute in a given application's environment. You can customize your environment variables by modifying your .env file. Here, NodeJS automatically loads environment variables into process.env to make them available to the application. 

> **Steps to add env variables**

Author: Got the point.

    - Sample .env file `.env.cep47.example` is already in the root folder 
3. Copy the content(a list of variables with values)) of sample .env file to the new file 
4. Replace the below parameters with your local usage settings
  - ***WASM_PATH***: Directory path for the generated .wasm file
  - ***CHAIN_NAME***: Network name that you are going to deploy the contract
  - ***NODE_ADDRESS***: [Address](./deploy#node_address) of the node 
  - ***EVENT_STREAM_ADDRESS***: [Address](./deploy#event_stream_address) of the event stream
  - ***MASTER_KEY_PAIR_PATH***: Path to the generated key pair
  - ***USER_KEY_PAIR_PATH***: Path to the generated key pair of the other party (In this case, it will be the same as your MASTER_KEY_PAIR_PATH )

:::note
It is mandatory to update the above list of parameters to align with your working environment.

*Note that you can directly include the environment values to the method by modifying [install()](https://github.com/casper-network/casper-contracts-js-clients/blob/b210261ba6b772a7cb25f62f2bdf00f0f0064ed5/e2e/cep47/install.ts#L52-L63) method in the install.ts file. On this tutorial we will use variables from .env.cep47 file for more clarity.*

> :::

Reviewer 2: I'm not sure this adds anything helpful - I'd be inclined to remove it.

### Arrange the JS packages and Resolve Dependencies
Run the npm package management command. This will manage and organize the packages and arrange the dependencies for the node.
```
npm install
```

### Building the JS Client Code 
Use the below command to build the JS client code and create the /dist directories with compiled files.
```
npm run dist
```

> After performing all the above steps you can start the deploy contract process and run the usage scenarios.

Reviewer 2: I'd be inclined to condense the two sections into one like:
```
### Building the JS Client
Run the following commands to install the dependencies and build the client:
[THREE BACKTICKS]
npm install
npm run dist
[THREE BACKTICKS]
```

## Deploying the Smart Contract to the Network
Run the below command to install your smart contract. This will use the selected node address on the Casper Testnet network. Command will execute the `./e2e/cep47/install.ts` file and you can check the outcome from the console.

```
npm run e2e:cep47:install
```

> If the execution proceeds without error, you should see a similar console output as below,

Reviewer 2: ```suggestion
If the execution succeeds, you should see output similar to the following:
```

```bash
... Contract installation deployHash: 0dcef7e7bddbc5a666aff1afbc03cf4797e3736c71fe05aee9944a26c4eeefab
... Contract installed successfully.
... Account Info:
{
  "_accountHash": "account-hash-179cd876d5c74317cce9c48d718a040e6e909063d7d786de0c5c6421a09fa803",
  "namedKeys": [
    {
      "name": "bdk_nft_contract_contract_hash",
      "key": "hash-a47d35d835a5fa8a1bcd55a4426dc14e21da9b876c1617742f18813737a4ece0"
    },
    {
      "name": "bdk_nft_contract_contract_hash_wrapped",
      "key": "uref-ff9b562d357d9a258acb2b3798f82c6ec5db49a8852e2e96b0ed4b1faf873206-007"
    },
    {
      "name": "contract_package_hash",
      "key": "hash-2468facdc9a6f324f8442584fd46d911e3ac9b434dfa79435567bf71f9b8bd23"
    }
  ],
  "mainPurse": "uref-a33e25cb1e6baa38e8306dba0492183c65cb41db3dbe8f69546868a4c0cfd0d9-007",
  "associatedKeys": [
    {
      "accountHash": "account-hash-179cd876d5c74317cce9c48d718a040e6e909063d7d786de0c5c6421a09fa803",
      "weight": 1
    }
  ],
  "actionThresholds": {
    "deployment": 1,
    "keyManagement": 1
  }
}
... Contract Hash: hash-a47d35d835a5fa8a1bcd55a4426dc14e21da9b876c1617742f18813737a4ece0

```

</details>

### Contract Installation Details

> This section clarifies how the contract deployment happens through the installation.js file.

Reviewer 2: There's no mention of `installation.js` before this.  Maybe we should provide the path to the file?

> First of all, the system takes the binary of the .wasm file and stores it in the *getBinary* constant.

Reviewer 2: ```suggestion
Firstly the client reads the contents of the .wasm file into the `getBinary` constant.
```

```javascript
export const getBinary = (pathToBinary: string) => {
  return new Uint8Array(fs.readFileSync(pathToBinary, null).buffer);
};
```

> Then, creates the token metadata fetch from the .env.cep47 file

Reviewer 2: ```suggestion
Then, it creates the token metadata fetched from the .env.cep47 file.
```

```javascript
const TOKEN_META = new Map(parseTokenMeta(process.env.TOKEN_META!));
```

> It also fetch the keys for signing from the .env.cep47 file.

Reviewer 2: ```suggestion
It also fetches the keys for signing from the .env.cep47 file.
```

```javascript
const KEYS = Keys.Ed25519.parseKeyFiles(
  `${MASTER_KEY_PAIR_PATH}/public_key.pem`,
  `${MASTER_KEY_PAIR_PATH}/secret_key.pem`
);
```

> Then, it fetches the node address and chain name of the network that you have planned to do the deploy.

Reviewer 2: ```suggestion
Then, it fetches the node address and chain name of the network being targeted.
```

```javascript
const test = async () => {
  const cep47 = new CEP47Client(
    NODE_ADDRESS!,
    CHAIN_NAME!
  ); 
```

> Run the installation, call *cep47.install()* function. This function will take WASM file path, token meta details, payment amount, public keys, and keys for multiple signing as parameters. The result will be stored in *installDeployHash* field.

Reviewer 2: ```suggestion
Next, it runs the installer by calling `cep47.install()`. This function takes the Wasm file path, token meta details, payment amount, public keys, and keys for multiple signing as parameters. The result is stored in `installDeployHash`.
```

```
It also fetches the keys for signing from the .env.cep47 file.

```javascript
const KEYS = Keys.Ed25519.parseKeyFiles(
  `${MASTER_KEY_PAIR_PATH}/public_key.pem`,
  `${MASTER_KEY_PAIR_PATH}/secret_key.pem`
);
```

Then, it fetches the node address and chain name of the network being targeted.

```javascript
const test = async () => {
  const cep47 = new CEP47Client(
    NODE_ADDRESS!,
    CHAIN_NAME!
  ); 
```
Next, it runs the installer by calling `cep47.install()`. This function takes the Wasm file path, token metadata, payment amount, and keys as parameters. The result is stored in `installDeployHash`.

```javascript
const installDeployHash = await cep47.install(
    getBinary(WASM_PATH!),
    {
      name: TOKEN_NAME!,
      contractName: CONTRACT_NAME!,
      symbol: TOKEN_SYMBOL!,
      meta: TOKEN_META
    },
    INSTALL_PAYMENT_AMOUNT!,
    KEYS.publicKey,
    [KEYS],
  );
```

Then the generated installation deploy hash is sent to the node address that you specified in the .env file. At this point, you can see the "... Contract installation deployHash: " message on the console output.

```javascript
const hash = await installDeployHash.send(NODE_ADDRESS!);
```

> After that, check if the deploy is successful and retrieve the account information using the node address and public key. Next, you can see the "Contract installed successfully.." message on the console.

Reviewer 2: ```suggestion
After that, check if the deploy is successful and retrieve the account information using the node address and public key. Next, you can see the "Contract installed successfully." message on the console.
```

# Contract Deployment
import useBaseUrl from '@docusaurus/useBaseUrl';

> Now that you have implemented the smart contract for CEP-47, it's time to deploy it to the blockchain. The deployment process for CEP-47 contract will be a bit different from ERC-20 contract. You will use a JavaScript client with in-built TypeScript (TS) classes to execute the CEP-47 contract deployment. The JS client also resides in a separate repository. You need to clone that repository to your machine and proceed with the steps to accomplish the deployment.

Reviewer: ```suggestion
Now that you have implemented the smart contract for CEP-47, it's time to deploy it to the blockchain. You will use a JavaScript client with in-built TypeScript (TS) classes to execute the CEP-47 contract deployment. The JS client also resides in a separate repository. Clone that repository to your machine and proceed with these deployment steps.
```

Reviewer: I recommend not mentioning ERC-20 here.

> Let's dive into CEP-47 contract deployment process.

Reviewer: ```suggestion
```

Reviewer: I think you can remove this line.

### Pre-requisites

> - Set up your machine as per the [prerequisites](/docs/workflow/setup)

Reviewer: This link doesn't seem to work for me, please double check it.

Author: It's working fine for me Iulia, Maybe someone else would have to check and confirm.

    - You will be using the `cep47-token.wasm` file from this repository for the deployment. This is the compiled implementation of the CEP-47 contract.
-  [Casper JavaScript Client](https://github.com/casper-network/casper-contracts-js-clients): This repository contains the files which execute the Javascript handle for the CEP-47 contract and other supporting classes to run the client. This contains the JavaScript handle for ERC-20 contract as well. 
    - You will be executing the `install.ts` file and `usage.ts` from this repository for the deployment

### Steps to Perform the Deployment

1. [Preparing the CEP-47 smart contract](./deploy#preparing-the-cep-47-smart-contract)
2. [Getting and preparing node addresses for the deployment](./deploy#getting-and-preparing-node-addresses-for-the-deployment)
3. [Preparing JS Client repository](./deploy#preparing-js-client-repository)
4. [Deploying the contract](./deploy#deploying-the-smart-contract-to-the-network)
5. [Deploying the entry points of the contract](../cep47/events)

## Preparing the CEP-47 Smart Contract

> Refer to the [contract preparation](../cep47/prepare) step to prepare the CEP-47 contract for deployment. This step will make the build environment, create the target location and compile the contract to a .wasm file.

Reviewer: This link doesn't work for me, please double check.

Author: Again, It is working for me and I will check it with someone else.

> - Output from this will be a `WASM file` (Eg:- cep47-token.wasm) which is later used by the JS compiler for the contract deployment

Reviewer: ```suggestion
- Output from this would be a `WASM file` (Eg: *cep47-token.wasm), which is later used by the JS client for contract deployment.
```

Reviewer: Do you mean JS client not JS compiler?

## Getting and Preparing Node Addresses for the Deployment

> Follow [getting an ip-address from a testnet peer](../erc20/deploy#getting-an-ip-address-from-a-testnet-peer) guide to get a node address for the deployment. Then you have to do some formatting to the address to include in the .env file.

Reviewer: This link doesn't seem to work. Can you check it, please? Also, capitalize IP and Testnet.

#### NODE_ADDRESS: 
- IP Address of the node
-  Use port '7777' instead of '35000' or any other and add "/rpc" to the end of the address 
    - (Eg:- 195.201.174.222:7777**/rpc**)
     
#### EVENT_STREAM_ADDRESS: 
- Address of the event stream
- To get this, replace 7777/rpc with --> '9999/events/main' phrase 
    - (Eg: 195.201.174.222:**9999/events/main**)

> ## Preparing JS Client Repository

Reviewer: ```suggestion
## Preparing the JS Client Repository
```

```
git clone https://github.com/casper-network/casper-contracts-js-clients.git
```

### Adding the Environment Variables 
Environment file (.env) contains the individual user environment variables required to execute in a given application's environment. You can customize your environment variables by modifying your .env file. Here, NodeJS automatically loads environment variables into process.env to make them available to the application. 

**Steps to add env variables**

Follow these steps to set up your environment variables to enable the CEP-47 smart contract deployment. 

1. Navigate to the cloned root folder (casper-contracts-js-clients)
2. Create a new environment file named `.env.cep47` 
    - Sample .env file `.env.cep47.example` is already in the root folder 
3. Copy the content(a list of variables with values)) of sample .env file to the new file 
4. Replace the below parameters with your local usage settings
  - ***WASM_PATH***: Directory path for the generated .wasm file
  - ***CHAIN_NAME***: Network name that you are going to deploy the contract
  - ***NODE_ADDRESS***: [Address](./deploy#node_address) of the node 

>   - ***EVENT_STREAM_ADDRESS***: [Address](./deploy#event_stream_address) of the event stream

Reviewer: Please check the links.

  - ***MASTER_KEY_PAIR_PATH***: Path to the generated key pair
  - ***USER_KEY_PAIR_PATH***: Path to the generated key pair of the other party (In this case, it will be the same as your MASTER_KEY_PAIR_PATH )

> :::note

Reviewer: The note doesn't display correctly on my end. Please double-check it.
